### PR TITLE
KV Store: StoreMessage and multipart upload implementation

### LINKED ISSUE #3242: KV multipart upload implementation

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #3242 and #3240

### CODE REVIEW

File: cmd/lakefs/cmd/run.go [hunk 1]
```diff
@@ -132,7 +134,18 @@ var runCmd = &cobra.Command{
...
-		multipartsTracker := multiparts.NewTracker(dbPool)
+		var multipartsTracker multiparts.Tracker
+
+		if dbParams.KVEnabled {
+			kvStore, err := kv.Open(ctx, dbParams.Type, dbParams.ConnectionString)
+			if err != nil {
+				logger.WithError(err).Fatal("failed to open KV store")
+			}
+			defer kvStore.Close()
+			multipartsTracker = multiparts.NewTracker(kv.StoreMessage{Store: kvStore})
```
Comment (itaiad200) [inline cmd/lakefs/cmd/run.go:145]: I know `migration` isn't part of this PR, but I would add 
```suggestion
			// TODO: migration required
			kvStore, err := kv.Open(ctx, dbParams.Type, dbParams.ConnectionString)
			if err != nil {
				logger.WithError(err).Fatal("failed to open KV store")
			}
			defer kvStore.Close()
			multipartsTracker = multiparts.NewTracker(kv.StoreMessage{Store: kvStore})
```

File: pkg/gateway/multiparts/multipart.proto [hunk 2]
```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
+option go_package = "github.com/treeverse/lakefs/gateway/multiparts";
+
+import "google/protobuf/timestamp.proto";
+
+package io.treeverse.lakefs.multiparts;
+
+message MultipartUploadData {
```
Comment (itaiad200) [inline pkg/gateway/multiparts/multipart.proto:8]: Please add documentation for the type and fields.

File: pkg/gateway/multiparts/tracker.go [hunk 3]
```diff
@@ -22,14 +26,36 @@ type MultipartUpload struct {
...
 
+func (m Metadata) Set(k, v string) {
+	m[strings.ToLower(k)] = v
+}
+
+func (m Metadata) Get(k string) string {
+	return m[strings.ToLower(k)]
+}
+func (m Metadata) Value() (driver.Value, error) {
+	return json.Marshal(m)
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker.go:37]: json?

File: pkg/gateway/multiparts/tracker.go [hunk 4]
```diff
@@ -22,14 +26,36 @@ type MultipartUpload struct {
...
+}
+
+func (m *Metadata) Scan(src interface{}) error {
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker.go:40]: Why not accept the concrete type `[]byte`?

File: pkg/gateway/multiparts/tracker.go [hunk 5]
```diff
@@ -38,21 +64,28 @@ var (
...
-		_, err := tx.Exec(`INSERT INTO gateway_multiparts (upload_id,path,creation_date,physical_address, metadata, content_type)
-			VALUES ($1, $2, $3, $4, $5, $6)`,
-			multipart.UploadID, multipart.Path, multipart.CreationDate, multipart.PhysicalAddress, multipart.Metadata, multipart.ContentType)
-		return nil, err
+	_, err := m.Get(ctx, multipart.UploadID)
+	if err == nil {
+		return ErrAlreadyExists
+	}
+
+	err = m.store.SetMsg(ctx, []byte(multipart.UploadID), &MultipartUploadData{
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker.go:82]: We need hierarchy for the keys in the store. Imagine creating a `Branch` and an `Upload` with the same ID.

File: pkg/gateway/multiparts/tracker.go [hunk 6]
```diff
@@ -61,59 +94,31 @@ func (m *tracker) Get(ctx context.Context, uploadID string) (*MultipartUpload, e
...
-	return m[strings.ToLower(k)]
-}
-func (m Metadata) Value() (driver.Value, error) {
-	return json.Marshal(m)
-}
-
-func (m *Metadata) Scan(src interface{}) error {
-	if src == nil {
-		return nil
+		return fmt.Errorf("failed on Get. key %s, %w", uploadID, err)
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker.go:120]: golang convention for wrapping errors:
```suggestion
		return fmt.Errorf("failed on Get. key (%s): %w", uploadID, err)
```

File: pkg/gateway/multiparts/tracker.go [hunk 7]
```diff
@@ -61,59 +94,31 @@ func (m *tracker) Get(ctx context.Context, uploadID string) (*MultipartUpload, e
...
 	}
-	data, ok := src.([]byte)
-	if !ok {
-		return ErrInvalidMetadataSrcFormat
-	}
-	return json.Unmarshal(data, m)
+	err := m.store.Delete(ctx, []byte(uploadID))
+	return err
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker.go:123]:
```suggestion
	return m.store.Delete(ctx, []byte(uploadID))
```

Comment (itaiad200) [reply pkg/gateway/multiparts/tracker.go:123]: Also - Does the store's `Delete` error on missing key? If so, no check is needed.

File: pkg/gateway/multiparts/multipart.pb.go [hunk 8]
```diff
@@ -0,0 +1,218 @@
+// Code generated by protoc-gen-go. DO NOT EDIT.
```
Comment (itaiad200) [inline pkg/gateway/multiparts/multipart.pb.go:1]: Should we commit this generated file?

File: pkg/kv/store_message.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
+package kv
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+type StoreMessage struct {
```
Comment (itaiad200) [inline pkg/kv/store_message.go:10]: Please add documentation for this Store. Not sure I understand what it does and why we need it.

File: pkg/kv/store_message.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	return s.Store.SetIf(ctx, key, val, curr)
+}
+
+func (s *StoreMessage) Delete(ctx context.Context, key []byte) error {
+	return s.Store.Delete(ctx, key)
+}
+
+// TODO: niro - implement when required
+// func (s *StoreMessage) Scan(ctx context.Context, start string) (Msgs, error) {
```
Comment (itaiad200) [inline pkg/kv/store_message.go:56]: It's better not to keep commented-out code.

File: pkg/kv/store_message_test.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+			"blue":  3,
+		},
+		TestList: []bool{true},
+	}
+
+	// SetIf fails nil
+	err = sm.SetIf(ctx, []byte(setModel.Name), m1, nil)
+	require.Error(t, kv.ErrPredicateFailed, err)
+
+	// SetIf fails nil
```
Comment (itaiad200) [inline pkg/kv/store_message_test.go:107]: wrong comment

File: pkg/gateway/multiparts/tracker_test.go [hunk 12]
```diff
@@ -73,6 +86,9 @@ func TestTracker_Get(t *testing.T) {
 				t.Errorf("Get() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
+			if !tt.wantErr {
+				got.CreationDate = got.CreationDate.UTC() // deepEqual workaround for time
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker_test.go:90]: So the time is the only thing that could fail?
What if you expect a different error and everything unexpectedly works as expected, except for the `CreationDate` which is flaky?

File: pkg/gateway/multiparts/tracker.go [hunk 13]
```diff
@@ -38,21 +64,28 @@ var (
 	ErrInvalidMetadataSrcFormat = errors.New("invalid metadata source format")
 )
 
-func NewTracker(adb db.Database) Tracker {
+func NewTracker(ms kv.StoreMessage) Tracker {
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker.go:67]: The name is generic, but it returns an implementation of the `Tracker`, no?

File: pkg/gateway/playback_test.go [hunk 14]
```diff
@@ -128,7 +128,7 @@ func getBasicHandler(t *testing.T, authService *simulator.PlayBackMockConf) (htt
 		DB:     conn,
 	})
 	testutil.MustDo(t, "build c", err)
-	multipartsTracker := multiparts.NewTracker(conn)
+	multipartsTracker := multiparts.NewDBTracker(conn)
```
Comment (itaiad200) [inline pkg/gateway/playback_test.go:131]: When will we test the new tracker here?

File: pkg/gateway/multiparts/tracker.go (hunk 3 above)
Comment (itaiad200) [reply pkg/gateway/multiparts/tracker.go:37]: OK it was just moved, never mind..

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Thanks for the first pkg implementation of the KV!
I have some concerns but overall looks great.

File: pkg/gateway/multiparts/multipart.pb.go (hunk 8 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/multipart.pb.go:1]: As in graveler - the pb files are committed to prevent repository from breaking

File: cmd/lakefs/cmd/run.go (hunk 1 above)
Comment (N-o-Z) [reply cmd/lakefs/cmd/run.go:145]: I'm not sure this will be the place where migration will happen, and since we don't know who might take this task, this TODO can easily be forgotten and left in the code. I'm confident we won't forget to implement the migration even without this TODO :)

File: pkg/gateway/multiparts/tracker.go (hunk 4 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:40]: This code was not touched - just moved in the file.
If we change this interface we will need also to touch the tracker usage which I was very careful not to disturb

File: pkg/gateway/multiparts/tracker.go (hunk 5 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:82]: I think this should be solved by making sure we create a unique hash (UUID / GUID) every time. Creating a conflicting BranchID / UploadID / RefID is a bug in my opinion

File: pkg/gateway/multiparts/tracker.go (hunk 7 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:123]: Delete doesn't error on missing key

File: pkg/gateway/multiparts/tracker_test.go (hunk 12 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker_test.go:90]: The CreationDate is not flaky it's just a different behavior between how it was saved originally in the DB and how protobuf's time object saves it.
protobuf's time object (timestamppb) doesn't save the timezone information so when read from the protobuf object we get the UTC time whereas the DB implementation when we read the timezone information, we get the time formatted using local timezone. The DeepEqual function has a "bug" where it compares two time objects without taking into account the timezone and this is simply a workaround to allow using DeepEqual and testing both implementations

File: pkg/gateway/multiparts/tracker.go (hunk 13 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:67]: This is the original name of the tracker and when we will be at feature complete this will also be the only implementation

File: pkg/gateway/playback_test.go (hunk 14 above)
Comment (N-o-Z) [reply pkg/gateway/playback_test.go:131]: Correct - fixed

File: pkg/gateway/multiparts/multipart.proto (hunk 2 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/multipart.proto:8]: We can add documentation for the message but I think the fields should be documented in the original struct

File: pkg/gateway/multiparts/tracker.go [hunk 15]
```diff
@@ -61,59 +101,30 @@ func (m *tracker) Get(ctx context.Context, uploadID string) (*MultipartUpload, e
...
 
 func (m *tracker) Delete(ctx context.Context, uploadID string) error {
 	if uploadID == "" {
 		return ErrInvalidUploadID
 	}
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		res, err := tx.Exec(`DELETE FROM gateway_multiparts WHERE upload_id = $1`, uploadID)
-		if err != nil {
-			return nil, err
+	if _, err := m.Get(ctx, uploadID); err != nil {
```
Comment (nopcoder) [inline pkg/gateway/multiparts/tracker.go:123]: This is good for debug, but not for performance. We can just call Delete on the store.

File: pkg/gateway/multiparts/tracker_test.go [hunk 16]
```diff
@@ -7,20 +7,33 @@ import (
...
+}
 
-	creationTime := time.Now().Round(time.Second) // round in order to remove the monotonic clock
+func TestDBTracker_Get(t *testing.T) {
+	tracker := testDBTracker(t)
+	testTrackerGet(t, tracker)
+}
+
+func testTrackerGet(t *testing.T, tracker multiparts.Tracker) {
+	t.Helper()
```
Comment (nopcoder) [inline pkg/gateway/multiparts/tracker_test.go:34]: would remove this one to get a better stack on failed test

File: pkg/gateway/multiparts/tracker_test.go [hunk 17]
```diff
@@ -7,20 +7,33 @@ import (
...
 	t.Helper()
 	conn, _ := testutil.GetDB(t, databaseURI)
-	return multiparts.NewTracker(conn)
+	return multiparts.NewDBTracker(conn)
 }
 
 func TestTracker_Get(t *testing.T) {
-	ctx := context.Background()
-	tracker := testTracker(t)
+	store := kvtest.MakeStoreByName("mem", "")(t, context.Background())
```
Comment (nopcoder) [inline pkg/gateway/multiparts/tracker_test.go:22]: 1. still prefer to test over postgres at this point
2. would split into two function to have code easy to read
3. why don't we have testTracker(t) to return a new tracker based on store - like we have for DB?

File: pkg/gateway/playback_test.go [hunk 18]
```diff
@@ -45,7 +48,17 @@ var (
...
 
+func TestGatewayRecordingDB(t *testing.T) {
+	viper.Set("database.kv_enabled", false)
+	testGatewayRecording(t)
+}
+
 func TestGatewayRecording(t *testing.T) {
+	viper.Set("database.kv_enabled", true)
+	testGatewayRecording(t)
+}
```
Comment (nopcoder) [inline pkg/gateway/playback_test.go:59]: 1. would not play with global viper configuration during test if possible - just passing an argument to testGatewayRecording if possible
2. check with @guy-har if this is still relevant or we can remove the recording parts

File: pkg/gateway/playback_test.go [hunk 19]
```diff
@@ -128,7 +140,16 @@ func getBasicHandler(t *testing.T, authService *simulator.PlayBackMockConf) (htt
 		DB:     conn,
 	})
 	testutil.MustDo(t, "build c", err)
-	multipartsTracker := multiparts.NewTracker(conn)
+
+	var multipartsTracker multiparts.Tracker
+	dbParams := conf.GetDatabaseParams()
```
Comment (nopcoder) [inline pkg/gateway/playback_test.go:145]: I don't think the test code need to use the configuration - it should setup the right services composition it would like to test

File: pkg/kv/store_message.go [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+		return err
+	}
+	return s.Store.SetIf(ctx, key, val, curr)
+}
+
+func (s *StoreMessage) Delete(ctx context.Context, key []byte) error {
+	return s.Store.Delete(ctx, key)
+}
+
+// TODO: niro - implement when required
```
Comment (nopcoder) [inline pkg/kv/store_message.go:56]: still relevant?

File: pkg/kv/store_message.go [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+		return err
+	}
+	return proto.Unmarshal(val, msg)
+}
+
+func (s *StoreMessage) SetMsg(ctx context.Context, key []byte, msg protoreflect.ProtoMessage) error {
+	val, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
```
Comment (nopcoder) [inline pkg/kv/store_message.go:27]: relevant for some places in this code - and probably in our storage too - the error should include more information like the key.

File: pkg/kv/store_message_test.go [hunk 22]
```diff
@@ -0,0 +1,211 @@
+package kv_test
```
Comment (nopcoder) [inline pkg/kv/store_message_test.go:1]: Like the non test package - we should keep the keys under specific prefix

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/gateway/multiparts/multipart.pb.go (hunk 8 above)
Comment (itaiad200) [reply pkg/gateway/multiparts/multipart.pb.go:1]: Not all generated code is committed, but I'm fine with keeping Graveler conventions.

File: pkg/gateway/multiparts/tracker_test.go (hunk 12 above)
Comment (itaiad200) [reply pkg/gateway/multiparts/tracker_test.go:90]: So why do it on the condition `!tt.wantErr`?

File: pkg/gateway/multiparts/tracker.go (hunk 13 above)
Comment (itaiad200) [reply pkg/gateway/multiparts/tracker.go:74]: Still think that `NewKVTracker` is a better name, but not blocking..

File: pkg/gateway/multiparts/tracker.go [hunk 23]
```diff
@@ -5,14 +5,25 @@ import (
...
+// MultipartUpload data structure
+// UploadID A unique identifier for the uploaded part
+// Path Multipart path in repository
+// CreationDate Creation date of the part
+// PhysicalAddress Physical address of the part in the storage
+// Metadata Additional metadata as required (by storage vendor etc.)
+// ContentType Original file's content-type
 type MultipartUpload struct {
 	UploadID        string    `db:"upload_id"`
 	Path            string    `db:"path"`
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker.go:29]: We usually document fields inline

```suggestion
// MultipartUpload represents an ongoing multipart upload to the blockstore
type MultipartUpload struct {
	// UploadID A unique identifier for the uploaded part
	UploadID        string    `db:"upload_id"`
	
	// Path Multipart path in repository
	Path            string    `db:"path"`
	
	...
```

File: pkg/gateway/multiparts/tracker_test.go (hunk 12 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker_test.go:90]: Because if we got an error then got == nil and this line will fail

File: pkg/gateway/multiparts/main_test.go [hunk 24]
```diff
@@ -8,6 +8,7 @@ import (
 
 	"github.com/ory/dockertest/v3"
 
+	_ "github.com/treeverse/lakefs/pkg/kv/mem"
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/multiparts/main_test.go:11]: I like the approach where unit tests use 'mem' and system tests use postgres. I think the only reason we used postgres in our unit tests is because we didn't have any other option.
We can take advantage of the new mem implementation and also save a lot of runtime. WDYT?

File: pkg/gateway/multiparts/tracker.go [hunk 25]
```diff
@@ -22,14 +26,36 @@ type MultipartUpload struct {
 	ContentType     string    `db:"content_type"`
 }
 
+func (m Metadata) Set(k, v string) {
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:35]: This is not new code - do you want me to modify it on this PR?

File: pkg/gateway/multiparts/tracker.go (hunk 15 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:122]: This is required since the current multiparts behavior relies on getting an error from delete in case the entry doesn't exist. I can change that but that will lead to behavior change

File: pkg/kv/store_message.go (hunk 20 above)
Comment (N-o-Z) [reply pkg/kv/store_message.go:56]: Yes sorry - this is a remark for implementing the Scan method - updated

File: pkg/gateway/playback_test.go (hunk 18 above)
Comment (N-o-Z) [reply pkg/gateway/playback_test.go:59]: Checked with @guy-har - the tests are not relevant anymore - removing it as part of this PR. (closes #2031)

File: pkg/gateway/playback_test.go (hunk 19 above)
Comment (N-o-Z) [reply pkg/gateway/playback_test.go:145]: Removed playback test

File: pkg/gateway/multiparts/tracker.go [hunk 26]
```diff
@@ -38,82 +72,70 @@ var (
...
 
-func NewTracker(adb db.Database) Tracker {
+func NewTracker(ms kv.StoreMessage) Tracker {
 	return &tracker{
-		db: adb,
+		store: ms,
+	}
+}
+
+func multipartFromProto(pb *MultipartUploadData) *MultipartUpload {
```
Comment (nopcoder) [inline pkg/gateway/multiparts/tracker.go:81]: This is great - nothing actionable - just a thought that maybe in the next iteration we can also support as part of the model's method. Similar to how the current DB can query for 'Scan' method to convert the structure to JSON. We can do the same to get from/to proto format.

**Review by nopcoder — APPROVED** [review]
Looks good thanks!, minor comments to consider.

File: pkg/gateway/main_test.go [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/spf13/viper"
+	"github.com/treeverse/lakefs/pkg/block"
+	"github.com/treeverse/lakefs/pkg/catalog"
+	"github.com/treeverse/lakefs/pkg/config"
+	"github.com/treeverse/lakefs/pkg/gateway"
+	"github.com/treeverse/lakefs/pkg/gateway/multiparts"
+	"github.com/treeverse/lakefs/pkg/gateway/simulator"
+	"github.com/treeverse/lakefs/pkg/kv"
+	"github.com/treeverse/lakefs/pkg/kv/kvtest"
+	_ "github.com/treeverse/lakefs/pkg/kv/mem"
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/main_test.go:22]: See my previous response:

> I like the approach where unit tests use 'mem' and system tests use postgres. I think the only reason we used postgres in our unit tests is because we didn't have any other option.
We can take advantage of the new mem implementation and also save a lot of runtime. WDYT?

File: pkg/gateway/main_test.go [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+	auth    simulator.GatewayAuthService
+	catalog catalog.Interface
+}
+
+const ReplayRepositoryName = "example"
+
+var (
+	pool         *dockertest.Pool
+	databaseURI  string
+	IdTranslator *testutil.UploadIDTranslator
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/main_test.go:38]: Used by getBasicHandler which is used by testPathWithTrailingSlash

File: pkg/gateway/main_test.go [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+	if err != nil {
+		log.Fatalf("Could not connect to Docker: %s", err)
+	}
+	databaseURI, closer = testutil.GetDBInstance(pool)
+	code := m.Run()
+	closer() // cleanup
+	os.Exit(code)
+}
+
+func setupTest(t *testing.T, method, target string, body io.Reader, kvEnabled bool) *http.Response {
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/main_test.go:117]: I'd rather not put a lot of thought into it as this is temporary code just until we complete the feature

File: pkg/gateway/multiparts/tracker.go [hunk 30]
```diff
@@ -38,82 +72,70 @@ var (
...
-		return nil
+		return fmt.Errorf("failed on Get. Key (%s): %w", uploadID, err)
 	}
-	data, ok := src.([]byte)
-	if !ok {
-		return ErrInvalidMetadataSrcFormat
-	}
-	return json.Unmarshal(data, m)
+	path := multipartPrefix + kv.PathDelimiter + uploadID
+	return m.store.Delete(ctx, path)
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:140]: I think it's better to do it in the store_message level

File: pkg/gateway/multiparts/db_tracker.go [hunk 31]
```diff
@@ -0,0 +1,69 @@
...
+		return ErrInvalidUploadID
+	}
+	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
+		res, err := tx.Exec(`DELETE FROM gateway_multiparts WHERE upload_id = $1`, uploadID)
+		if err != nil {
+			return nil, err
+		}
+		affected := res.RowsAffected()
+		if affected != 1 {
+			return nil, ErrMultipartUploadNotFound
```
Comment (itaiad200) [inline pkg/gateway/multiparts/db_tracker.go:64]: Isn't the KV behave differently in this case?

File: pkg/gateway/multiparts/tracker.go (hunk 15 above)
Comment (itaiad200) [reply pkg/gateway/multiparts/tracker.go:132]: Nice catch, I found it confusing in the original implementation but I guess this is how S3 behaves 🤷

File: pkg/gateway/multiparts/tracker.go [hunk 32]
```diff
@@ -38,82 +72,70 @@ var (
...
-	if src == nil {
-		return nil
+		return fmt.Errorf("failed on Get. Key (%s): %w", uploadID, err)
 	}
-	data, ok := src.([]byte)
-	if !ok {
-		return ErrInvalidMetadataSrcFormat
-	}
-	return json.Unmarshal(data, m)
+	path := multipartPrefix + kv.PathDelimiter + uploadID
```
Comment (itaiad200) [inline pkg/gateway/multiparts/tracker.go]: Not part of this PR - I think that one of the utilities of the KV package should be concatenating paths. We messed up with it so many times with the block adapters..

File: pkg/gateway/multiparts/tracker.go [hunk 33]
```diff
@@ -5,21 +5,55 @@ import (
...
+	"fmt"
 	"strings"
 	"time"
 
-	"github.com/treeverse/lakefs/pkg/db"
+	"github.com/treeverse/lakefs/pkg/kv"
+	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
+const multipartPrefix = "multipart"
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:16]: Implementing NewPath on kv

File: pkg/gateway/multiparts/tracker_test.go (hunk 12 above)
Comment (itaiad200) [reply pkg/gateway/multiparts/tracker_test.go:89]: Got it. The condition should actually be `(err == nil)` but the previous check makes it the same as `!tt.wantErr`. Consider changing to `(err == nil)` for clarity.

File: pkg/kv/store_message.go [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+
+// StoreMessage protobuf generic implementation for kv.Store interface applicable for all data models
+type StoreMessage struct {
+	Store Store
+}
+
+func (s *StoreMessage) GetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
+	val, err := s.Store.Get(ctx, []byte(path))
+	if err != nil {
+		return fmt.Errorf("failed on Get. Path (%s): %w", path, err)
```
Comment (itaiad200) [inline pkg/kv/store_message.go]: Here and elsewhere

I think that the way errors are wrapped here may lead to long error lines with repetitive information, imagine the following possible error:
```
failed on Get. Path (this/is/some/very/long/repetitive/path): failed on Get. Path (this/is/some/very/long/repetitive/path): failed on Get. Path (this/is/some/very/long/repetitive/path): postgres failed to connect. Path (this/is/some/very/long/repetitive/path): tcp: no such host
```
Now it's repetitive because there are several calls to Get with path (gateway, storeMessage, KVStore and probably others I missed.)

I suggest the following (I think it's the golang approach)
```suggestion
		return fmt.Errorf("getting from store: %w", err)
```

If the path is relevant for the error, it should be added at the very start of the chain where the path has the most meaning.

**Review by itaiad200 — APPROVED** [review]
Few minor comments 
LGTM

File: pkg/gateway/main_test.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/main_test.go:117]: This code won't be removed (checking with @guy-har if the test using it can be removed altogether) but the kvEnabled flag will no longer be relevant once we are in feature complete and we will only be using and testing one tracker implementation

File: pkg/gateway/main_test.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/main_test.go:22]: Since this is a unit test and we are testing the tracker functionality and not the KV functionality I believe the underlying KV implementation is irrelevant. In any case - the postgres implementation should be tested by it's own in a dedicated unit test and will be tested extensively as part of the system tests

File: pkg/gateway/multiparts/db_tracker.go (hunk 31 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/db_tracker.go:64]: It doesn't since message_store Delete I explicitly check if  the doesn't exist and return ErrMultipartUploadNotFound immediately in this case to simulate this behavior. This is related to @nopcoder's comment as well:

> This is required since the current multiparts implementation relies on getting an error from delete in case the entry doesn't exist. I can change that but that will lead to behavior change

File: pkg/gateway/multiparts/tracker.go (hunk 32 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go]: Done

File: pkg/gateway/multiparts/tracker_test.go (hunk 12 above)
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker_test.go:89]: Done

**Comment by arielshaqed** [comment]
Note that #3303 will soon remove almost all of the gateway test.  IIWY, I would minimize the changes there.  And I'll be happy to help resolve the coming conflicts in any way that I can.

**Comment by N-o-Z** [comment]
> Note that #3303 will soon remove almost all of the gateway test. IIWY, I would minimize the changes there. And I'll be happy to help resolve the coming conflicts in any way that I can.

Perfect @arielshaqed thanks for the heads up. So I think I'll just wait for this ticket to merge and rebase my changes - since we still need to modify GetBasicHandler multipartsTracker initialization

**Comment by N-o-Z** [comment]
Rebased over master after #3303 changes. Added NewPath function.

File: pkg/gateway/multiparts/tracker.go (hunk 15 above)
Comment (nopcoder) [reply pkg/gateway/multiparts/tracker.go:107]: I assume we will have to verify if the S3 protocol does it too - if it is not the case we will change the behviour as there is no way to ensure that by calling Get a second before Delete.

File: pkg/gateway/multiparts/tracker.go [hunk 35]
```diff
@@ -38,82 +46,70 @@ var (
...
 
 func (m *tracker) Delete(ctx context.Context, uploadID string) error {
 	if uploadID == "" {
 		return ErrInvalidUploadID
 	}
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		res, err := tx.Exec(`DELETE FROM gateway_multiparts WHERE upload_id = $1`, uploadID)
-		if err != nil {
-			return nil, err
+	if _, err := m.Get(ctx, uploadID); err != nil {
```
Comment (nopcoder) [inline pkg/gateway/multiparts/tracker.go:107]: I would not implement Delete in the tracker level, using Get in the tracker level - until there is another lower implementation of 'get'.
This is because usually the logging, BI, etc will be implemented at that level.
In this above case - I'll use the storage Get and not the upper level one.

File: pkg/kv/store.go [hunk 36]
```diff
@@ -20,6 +23,14 @@ var (
 	ErrUnknownDriver       = errors.New("unknown driver")
 )
 
+func NewPath(p ...string) string {
+	const minPathLength = 2
+	if len(p) < minPathLength {
+		panic("invalid path length")
+	}
```
Comment (nopcoder) [inline pkg/kv/store.go:30]: Why? keep it simple and just enable us to build a single word path - which will not include any sep.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/gateway/multiparts/tracker.go [hunk 37]
```diff
@@ -38,82 +46,70 @@ var (
...
-			VALUES ($1, $2, $3, $4, $5, $6)`,
-			multipart.UploadID, multipart.Path, multipart.CreationDate, multipart.PhysicalAddress, multipart.Metadata, multipart.ContentType)
-		return nil, err
-	})
+	_, err := m.Get(ctx, multipart.UploadID)
+	if err == nil {
+		return ErrAlreadyExists
+	}
+	path := kv.NewPath(multipartPrefix, multipart.UploadID)
+	err = m.store.SetMsg(ctx, path, protoFromMultipart(&multipart))
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go:86]: @nopcoder - If we use SetIf in this instance - we will get an ambiguous error on ErrPredicateFailed which we will need to interpret as ErrAlreadyExists. We will also not save on any API calls since SetIf uses Get as well

File: pkg/kv/store_message.go [hunk 38]
```diff
@@ -37,14 +37,14 @@ func (s *StoreMessage) SetIf(ctx context.Context, path string, msg protoreflect.
...
-		return fmt.Errorf("failed on Unmarshal (path: %s) (curr msg: %s): %w", path, currMsg, err)
+		return fmt.Errorf("failed on Unmarshal (path: %s) %w", path, err)
 	}
 	if !proto.Equal(pred, currMsg) {
 		return fmt.Errorf("failed on predicate. Path (%s): %w", path, ErrPredicateFailed)
 	}
 	val, err := proto.Marshal(msg)
 	if err != nil {
-		return fmt.Errorf("failed on Marshal (path: %s) (msg: %s): %w", path, msg, err)
+		return fmt.Errorf("failed on Marshal (path: %s): %w", msg, err)
```
Comment (nopcoder) [inline pkg/kv/store_message.go:47]: path

File: pkg/gateway/multiparts/tracker.go (hunk 33 above)
[the remark this answers is not in the document]
Comment (nopcoder) [reply pkg/gateway/multiparts/tracker.go:12]: Nice - maybe a different name for the function FormatPath or something like that. The name `NewPath` suggests it will return a type Path.

File: pkg/gateway/multiparts/tracker.go (hunk 37 above)
[the remark this answers is not in the document]
Comment (nopcoder) [reply pkg/gateway/multiparts/tracker.go:86]: Understood about the error type migration - but it is in place to handle the 'transaction' part of setting a new value, just in case it doesn't exist.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/gateway/multiparts/tracker.go (hunk 37 above)
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/gateway/multiparts/tracker.go]: I've modified the SetIf function in store_message. Please have another look

**Review by nopcoder — APPROVED** [review]
LGTM!
